pxd_fastpath.c

```diff
@@ -435,42 +474,45 @@ static void pxd_complete_io(struct bio* bio, int error)
 	struct pxd_io_tracker *iot = container_of(bio, struct pxd_io_tracker, clone);
 	struct pxd_device *pxd_dev = bio->bi_private;
 	struct pxd_io_tracker *head = iot->head;
-	bool dofree = true;
+	unsigned int flags = get_op_flags(bio);
+	int blkrc;
 
-	fput(iot->file);
 	BUG_ON(iot->magic != PXD_IOT_MAGIC);
 	BUG_ON(head->magic != PXD_IOT_MAGIC);
 	BUG_ON(pxd_dev->magic != PXD_DEV_MAGIC);
-	if (!atomic_dec_and_test(&head->active)) {
-		// not all responses have come back
-		// but update head status if this is a failure
+
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,13,0) ||  \
     (LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0) && \
      defined(bvec_iter_sectors))
-		if (bio->bi_status) {
-			atomic_inc(&head->fails);
-		}
+		blkrc = blk_status_to_errno(bio->bi_status);
 #elif LINUX_VERSION_CODE >= KERNEL_VERSION(4,3,0)
-		if (bio->bi_error) {
-			atomic_inc(&head->fails);
-		}
+		blkrc = bio->bi_error;
 #else
-		if (error) {
-			atomic_inc(&head->fails);
-		}
+		blkrc = error;
 #endif
-		pxd_printk("pxd_complete_io for bio %px (pxd %px) with head %px active %d error %d early return\n",
-			bio, pxd_dev, head, atomic_read(&head->active), atomic_read(&head->fails));
-
-		return;
-	}
 
-	pxd_io_printk("%s: dev m %d g %lld %s at %ld len %d bytes %d pages "
-			"flags 0x%lx\n", __func__,
+	if (blkrc != 0) {
+		printk_ratelimited("FAILED IO %s (err=%d): dev m %d g %lld %s at %ld len %d bytes %d pages "
+				"flags 0x%lx\n", __func__, blkrc,
 			pxd_dev->minor, pxd_dev->dev_id,
 			bio_data_dir(bio) == WRITE ? "wr" : "rd",
 			BIO_SECTOR(bio) * SECTOR_SIZE, BIO_SIZE(bio),
-			bio->bi_vcnt, bio->bi_flags);
+			bio->bi_vcnt, (long unsigned int)flags);
+	}
+
+	fput(iot->file);
+	iot->status = blkrc;
+	if (!atomic_dec_and_test(&head->active)) {
+		// not all responses have come back
+		return;
+	}
+
+	// final reconciled status
+	blkrc = reconcile_io_status(head);
+
+	// debug condition for force fail
+	if (pxd_dev->fp.force_fail) blkrc = -EIO;
```

Comment:
Don't think the force_fail really needed? Can you not simulate this by other means for example, pulling a disk out from under the btrfs mountpoint?

Reply:
This is a debug interface to test, this can go away once all pieces are found stable.. Let it remain until independent testing completes. I will remove it then.